Repair 0.22 China

knoop7 opened the pull request:
Hello, some of the modifications have been made, you can not choose to merge, I hope you will check it

ReneNulschDE commented:
Hi, can you summarize the real problems and the proposed changes. This PR is not mergable. I propose that you create a branch from version 0.22 in your repo, make the needed changes and share the branch name, so that I can check it.

knoop7 commented:
### 1. china api endpoint update
**File:** `const.py`  
```python
REST_API_BASE_CHINA = "https://bff.mobileapp.mercedes-benz.com.cn"
```

### 2. http request headers for china region
**File:** `webapi.py`  
```python
if self._region == REGION_CHINA:
    kwargs["headers"] = {
        "Authorization": f"Bearer {token['access_token']}",
        "X-SessionId": self.session_id,
        "X-TrackingId": str(uuid.uuid4()).upper(),
        "X-Platform-Id": "ios",
        "X-Authmode": "CIAMNG",
        "X-Application-Id": "mbapp",
        "Cache-Control": "no-store",
        "User-Agent": "MyStarCN/1.54.1 (com.daimler.ris.mercedesme.cn.ios; build:2744; iOS 18.4.0) Alamofire/5.9.1",
        "Accept": "application/json",
        "Accept-Language": "zh-Hans-TW;q=1.0, en-TW;q=0.9",
        "Content-Type": "application/json; charset=UTF-8",
        "Connection": "keep-alive",
        "Accept-Encoding": "br;q=1.0, gzip;q=0.9, deflate;q=0.8",
    }
```

### 3. api endpoint version change
**File:** `webapi.py`  
```python
async def get_user_info(self):
    return await self._request("get", "/v1/vehicles")
```

### 4. geofencing error fix
**File:** `client.py`  
```python
geofencing_violotions = await self.webapi.get_car_geofencing_violations(car.finorvin)
if geofencing_violotions and len(geofencing_violotions) > 0:
    car.geofence_events.last_event_type = CarAttribute(
        geofencing_violotions[-1].get("type"),
        "VALID",
    )
```

### 5. sensor state inversion
**File:** `sensor.py`  
```python
elif self._internal_name == "windowStatusOverall":
    return not self._state
```  If the modification is not successful, you can ignore it. 🥲

## Benefits of Changes
- Fixed HTTP 418 errors by updating China API endpoints and headers
- Resolved 404 errors by changing API version from v2 to v1
- Prevented crashes by adding null check for geofencing violations
- Corrected window sensor state to show "off" when windows are closed

## Deployment Instructions
1. Replace the modified files in your custom_components/mbapi2020 directory
2. Restart Home Assistant
3. Verify sensor states in developer tools

knoop7 commented:
Hello, I hope you don't mind, because 0.27 my development environment can't run normally, so follow the 0.22 path to commit, just for your reference, according to the modification found that there will be no error and normal operation 
<img width="1028" height="430" alt="image" src="https://github.com/user-attachments/assets/6ed3a951-b256-4ab5-a435-8fee0e1a5195" />

knoop7 commented:
> Thanks, let me try to create a branch v.0.22.1 with your changes and you can check it. If this works I can publish a v.0.22.1 for China only. OK for you?


Thank you so much, yes, I just hope you can refer to it, due to the particularity of the country, so the branch is completely different, and the region may be updated and adjusted accordingly in the future

ReneNulschDE commented:
ok, let me check this - but give me some time, I need to test this workflow and plan some time for this  ;-)

ReneNulschDE commented:
I asked the HACS team if they have a solution for it: https://github.com/hacs/integration/issues/4706

knoop7 commented:
> I made some tests with HACS. :-( It is not possible to do a release from a different branch. See my test repo https://github.com/ReneNulschDE/hacstest-china - when you add this as a custom repo to hacs, you could download v0.22 and v0.27 but not v0.22.1 (0.22.1 is released from the special "china"-branch...
> 
> One option would be: You create a repo with the latest changes and I add note that the China solution is in your repo... What do you think about this? Do you have other ideas?

Thank you, the test was successful, thank you for your efforts, your integration is very useful

knoop7 commented:
> I asked the HACS team if they have a solution for it: [hacs/integration#4706](https://github.com/hacs/integration/issues/4706)

Ok, you can continue to update with the new version, I still have a chance to try to follow your new version and try to blend as little as possible without breaking your main story, you just need to review it when the time comes
